### ISSUE TITLE
Complience with CF-1.7

### ISSUE THREAD

**Issue by paalge** [acolite/acolite#5]
This is a continuation of https://odnature.naturalsciences.be/remsem/acolite-forum/viewtopic.php?f=4&t=261&p=1005#p1005
with the intention of creating a pull request. 

I have looked at for instance: https://github.com/acolite/acolite/blob/8729d0b101181742aa617858cdcd8e6867e7a4e0/acolite/acolite/acolite_l2w.py#L159 

which is not a [recognised standard name](https://cfconventions.org/Data/cf-standard-names/76/build/cf-standard-name-table.html), which I guess should be surface_ratio_of_upwelling_radiance_emerging_from_sea_water_to_downwelling_radiative_flux_in_air

I can go through and suggest these and other names in a pull request if that is the best approach?

**Comment by acolite** [comment]
Hi Pål

That's a fine approach for me. For the files to be CF compliant, are there other changes than fixing the "standard_name" attributes that need to be done?

Or would removing the "standard_name" attribute also work? It does bother me a little bit that there is "sea" in the standard name for Rrs for example. :-)

Quinten

**Comment by paalge** [comment]
Variables should all have _standard_name_ , _long_name_, _units_
This includes x,y, lat, lon 
It is also common to add _time_ as the first axis in the variables, (i.e. (time, x, y)), though I suspect that this requires some significant rewriting of the code.
Another warning was the use of global parameters starting with a number _(WARN: (2.3): Global attribute 1_wave: Attribute names should begin with a letter and be composed of letters, digits and underscores)_

Generally one wants _standard_name_ as it explains what the variable is in a controlled manner (i.e. the CF standard names) making the use of the dataset easier especially in automated systems. I would advise to use them where possible, as [FAIR](https://www.go-fair.org/fair-principles/) datasets is becoming more and more a requirement when publishing data. 

Then to which _standard_name_ to choose I might not be the right one to choose them, as I don't have in depth knowledge on all the parameters generated by acolite. If you are bothered by "sea" in Rrs, then it is most likely not the correct name. 
Not all parameters have a "standard_name" in the table, but there are guidlines for creating them: http://cfconventions.org/Data/cf-standard-names/docs/guidelines.html

**Comment by acolite** [comment]
Thanks for your comments! The "sea" makes sense as Rrs has been used for a long time in "ocean colour" but we are looking a lot at inland waters and rivers now, which don't have "sea water". :-)

I think I have a flexible approach to the writing of the attributes. Rather than hard coding them in any of the l1_convert, acolite_l2r or acolite_l2w routines, I will make a lookup table that has the variable, standard_name, long_name, and units. The NetCDF writer can then access this table to get the CF attributes.

Is it worth changing the NetCDF dimensions from x,y to lon, lat rather than assigning x,y dimensions to the lon and lat variables? https://github.com/acolite/acolite/blob/8729d0b101181742aa617858cdcd8e6867e7a4e0/acolite/output/nc_write.py#L72-L74

Quinten

**Comment by paalge** [comment]
I agree that it is no longer the case and a new name should be used. Maybe worth also sending this as an input to the standard names governance.

Sounds like a good idea. I was looking through the code and have started some suggetions, but it was not that easy to know where to put them. 
An example of the use on sentinel-2 data can be seen [here: ](http://nbstds.met.no/thredds/dodsC/NBS/S2A/2019/04/09/S2A_MSIL1C_20190409T105621_N0207_R094_T33WWR_20190409T130615.nc.html)

As far as I understand x,y (m) is the standard way of doing it, with lat,lon as you have it (see example above)

**Comment by acolite** [comment]
OK! 

I suggest we make a CSV file with all parameters, and I will implement a way to access the file and the proper attributes when the files are written. For example "variable_cf_attributes.csv" with a header line, and then a line per variable, including some with a wild card that can have a wavelength dependency:

```
variable,standard_name,long_name,units
Rrs_*,surface_ratio_of_upwelling_radiance_emerging_from_sea_water_to_downwelling_radiative_flux_in_air,Remote sensing reflectance,sr-1
TUR_Nechad_*,sea_water_turbidity,water turbidity in NTU,1
```

**Comment by paalge** [comment]
I was wondering if a csv file was the best for this? 
Maybe a file with python dictionaries, or a yaml file would be better? Then if you need to add more or fewer fields later, it should be easier.

**Comment by acolite** [comment]
Hi Pål

I added CF attributes in the last commit, let me know if you find any issues! 

https://github.com/acolite/acolite/commit/83e5bd6ba3a3be18de42314fec575dd3b9659998

Individual parameter configuration is in config/parameter_cf_attributes.json as a list of dicts, but could be changed.

Quinten

**Comment by paalge** [comment]
I looks mostly good, though if you didn't like the sea_water part, I would change it to water (as i agree we are moving to freshwater as well)

The units for x and y should also be there

One thing I could not see is the insertion of the `Conventions='CF-1.7'` global attribute.

**Comment by acolite** [comment]
Hi Pål

Thanks for your comments, I will add the convention to the global attributes. 

Do you mean the UTM x/y easting/northing in metres, or the x/y "pixel" coordinates/dimensions in the output image? I am not sure on how to add the latter.

Quinten

**Comment by paalge** [comment]
I mean the variables x/y (might be UTM, they are currently not named) in the netcdf files.
There it would be good to add the projection information in some form.
I know that the MET office in Norway uses something like this:

```        
        int x(x) ;
                x:units = "m" ;
                x:standard_name = "projection_x_coordinate" ;
                x:_ChunkSizes = 10980 ;
        int y(y) ;
                y:standard_name = "projection_y_coordinate" ;
                y:units = "m" ;
                y:_ChunkSizes = 10980 ;
        int projection ;
                projection:longitude_of_central_meridian = 15. ;
                projection:grid_mapping_name = "transverse_mercator" ;
                projection:semi_minor_axis = 6356752.31424518 ;
                projection:false_easting = 500000. ;
                projection:semi_major_axis = 6378137. ;
                projection:scale_factor_at_central_meridian = 0.9996 ;
                projection:latitude_of_projection_origin = 0. ;
                projection:proj4_string = "+proj=utm +zone=33 +datum=WGS84 +units=m +no_defs " ;
```

for processed data:
http://nbstds.met.no/thredds/dodsC/NBS/S2A/2019/04/09/S2A_MSIL1C_20190409T105621_N0207_R094_T33WWR_20190409T130615.nc.html
